### ISSUE TITLE
Contributing with Janitor

### ISSUE THREAD

**Issue by EtienneWan** [yuzu-emu/yuzu#35]
Hello,
We are hosting a platform named Janitor (https://janitor.technology) to bootstrap development containers for open-source projects.
We are interested into hosting the project and help you contribute faster.
I looked at your build instructions and it seems relatively simple to create a Dockerfile for our platform to clone the project, install all dependencies and build yuzu.
We feature a cloud IDE (powered by our own instances of Cloud9 IDE), Ubuntu containers running on bare metal servers, a desktop preview with noVNC and debug tools working out-of-the-box.

If you are interested in this idea, let's continue the discussion here for specific needs for your Dockerfile or meta questions. If you are not, feel free to close this issue.

Happy coding,
EtienneWan

**Comment by j-selby** [comment]
Hi! Thanks for the generous offer!

I'll let some other devs chime in, but it looks like it would work quite well.

Getting a Dockerfile going is trivial - we already use Docker in our normal build pipeline for Linux, so that shouldn't be too hard of a hurdle. Here is one which gets us most of the way: https://gist.github.com/j-selby/ae15ac5dafca1d8d56971e5007912daf

**Comment by EtienneWan** [comment]
Pretty straight forward dockerfile indeed.
I think you should host the dockerfile in a repo in this organisation, and push the image once it has been built to hub.docker.com. You can tag it like you want.
Once this is done, we will pull the image on our server and you will be able to spawn containers for yuzu.

For the dockerfile, here are my edits :
```
FROM janx/ubuntu-dev
MAINTAINER Selby "jselby@jselby.net"

RUN sudo apt-get update -q \
 && sudo apt-get install -y libqt5opengl5-dev libsdl2-dev libssl-dev qtbase5-dev

# We use a newer version of CMake then is provided by Ubuntu
RUN wget https://cmake.org/files/v3.10/cmake-3.10.1-Linux-x86_64.sh \
 && mkdir /home/user/cmake \
 && sh cmake-3.10.1-Linux-x86_64.sh --exclude-subdir --prefix=/home/user/cmake --skip-license
ENV PATH="/home/user/cmake/bin:${PATH}"

RUN git clone https://github.com/yuzu-emu/yuzu /home/user/yuzu --recursive \
 && mkdir /home/user/yuzu/build
WORKDIR /home/user/yuzu/build

# Build
RUN cmake .. -DYUZU_BUILD_UNICORN=ON -DCMAKE_BUILD_TYPE=Release \
 && make -j4

# Configure Cloud9 to use yuzu source directory as workspace (-w).
RUN sudo sed -i "s/-w \/home\/user/-w \/home\/user\/yuzu\/" /etc/supervisord.conf
```

Maybe we should add a `janitor.json` file, but we'd better ask @JanKeromnes for that. :wink:

**Comment by jankeromnes** [comment]
Super cool initiative! I see that most of the work to add Yuzu to [Janitor](https://janitor.technology) is already done. 😄 

Having the Dockerfile somewhere in this repo, and hooking the repo up as a Docker Hub automated build (takes like 3-5 clicks), would allow Docker Hub to detect any new commits to https://github.com/yuzu-emu/yuzu/ and rebuild the developer image automatically, keeping it up-to-date with no effort from anyone.

Adding a `janitor.json` configuration file is pretty easy too, and helps to (1) expose web ports via Janitor (but if your project doesn't need to expose web servers to its developers you can skip this), and more importantly (2) populate the IDE's `Run` button with useful project commands.

Adding a `janitor.json` to this development Dockerfile can be done [like so](https://github.com/Chocobozzz/PeerTube/commit/b270ad07f13fa4125fbade6208596071329bc92f). We just need to know a few useful commands that Yuzu developers use all the time (e.g. build, run, test, send patch, etc).

**Comment by jankeromnes** [comment]
P.S.

- `ENV PATH="/home/user/cmake/bin:${PATH}"` this needs to be `echo`'d into `.bashrc` as well, because `ENV` instructions don't transpire into user SSH sessions once the image is built. See [this example](https://github.com/JanitorTechnology/dockerfiles/blob/905b6d2a23c5be9fb019aa3ad2ba05737e9bd1ce/chromium/chromium.dockerfile#L6-L8).
- `make -j4` I think you could bump this to `-j18` (event though Docker Hub's current infra doesn't have this, I believe it's more future-proof, and shouldn't impact current automated build times too much)

**Comment by chris062689** [comment]
Hello!

Thank you for reaching out to us. We'd love to schedule a phone call to discuss this further. 

Let us know when you are available!
yuzu@yuzu-emu.org
